### ISSUE TITLE
New Steam UI notifications steal focus on Wayland

### ISSUE THREAD

**Issue by ErikReider** [ValveSoftware/steam-for-linux#9458]
#### Your system information

* Steam client version (build number or date): 1683078372
* Distribution (e.g. Ubuntu): Fedora 38
* Opted into Steam client beta?: Yes
* Have you checked for system updates?: Yes

#### Steps for reproducing this issue:

1. Focus any input field in any application
2. Wait until notification pops up
3. The notification steals the focus

I'm definitely not an X11 ATOM expert but when running a `xprop` on the notifications, the `WM_TAKE_FOCUS` ATOM is in-use.

On the topic of notifications, providing a toggle in the settings to use the system notification daemon would be nice. Would make it easier to dismiss and control where notifications appear :)

**Comment by ardishco-the-great** [comment]
This bug has still not been fixed in the recent stable release and it is happening as I type this comment.

I think for most Wayland WMs there's some way to make it so that it never get focused and rendered as an overlay layer.

**Comment by purplebar0** [comment]
Still present.

According to [window manager logs](https://steamcommunity.com/groups/SteamClientBeta/discussions/0/3818536354781079260/#c6717729343878367441), Steam spawns a stray window with title = "None" and kills it 20 ms later.

This is extremely annoying, it steals focus and shifts all windows on any tiling window manager.

I attempted to make floating window and group rules as a workaround, but for some reason, Qtile could not match name="None" even by itself.

**Comment by JacobSvenningsen** [comment]
It also happens on X with XFWM (stacking window manager), although it sometimes just stops happening too. 
Haven't found a reliable way to reproduce it.

Most often it occurs after launching Steam. As @purplebar0 pointed out, the notification that's disappearing and spawning said error in the log, steals the focus temporarily.
However, even though the previous focused window receives focus again, it isn't actually received "completely".
What I mean by this, is despite you thinking the window has focus, you won't be able to type anything in said window without first alt-tabbing to a Steam-owned process, then tabbing back again. This can be annoying if you're working in a text-editor for instance, since suddenly you won't be able to input anything until you've tabbed in and out. And this is necessary every time a notification disappears

This started happening after the beta got pushed to the stable branch. It wasn't always an issue with the beta.
Or maybe it was and I just got lucky for not having encountered it, seeing it doesn't always happen

**Comment by purplebar0** [comment]
> However, even though the previous focused window receives focus again, it isn't actually received "completely".

> What I mean by this, is despite you thinking the window has focus, you won't be able to type anything in said window without first alt-tabbing to a Steam-owned process, then tabbing back again.

This is part of a different issue which can be reproduced by switching back and forth from the Chat window, where the text field remains active after switching from the Chat window.

**Comment by purplebar0** [comment]
Update: This does not only affect notifications, but _any_ windows that are to be closed. Tested with the Steam Settings window.

On Qtile + Picom, it is super annoying even on Max layout because smooth window transitions have to display that stray invisible window and it ends up being a flashing window.

**Comment by lostgoat** [comment]
If you run with this environment variable set do those windows go away for you:
```
STEAM_DISABLE_BROWSER_SHUTDOWN_WORKAROUND=1 steam
```

**Comment by purplebar0** [comment]
> If you run with this environment variable set do those windows go away for you:
> 
> ```
> STEAM_DISABLE_BROWSER_SHUTDOWN_WORKAROUND=1 steam
> ```

Yes! The windows don't instantly close, but at least the annoying side-effect of the workaround isn't present any longer.

**Comment by lostgoat** [comment]
Just to confirm, does the env fix both the flashing and the focus issue, or just one of them?

**Comment by Sid127** [comment]
Can't confirm. Never had the flashing issue, and the env var doesn't fix the focus issue either. Using sway, fwiw

**Comment by purplebar0** [comment]
> Just to confirm, does the env fix both the flashing and the focus issue, or just one of them?

I can confirm that it fixes both in my case (Qtile WM).

**Comment by purplebar0** [comment]
Using an AMD GPU, I can confirm that today's beta disables the workaround without requiring the environment variable.

**Comment by ErikReider** [comment]
> Using an AMD GPU, I can confirm that today's beta disables the workaround without requiring the environment variable.

Still an issue for me on sway. Haven't tried the env variable though

**Comment by Bettehem** [comment]
I'm using Sway too, the issue is not fixed for me in the latest beta and the environment variable doesn't seem to help either.

**Comment by InventorXtreme** [comment]
I am on bspwm on X and still have this issue.

**Comment by DrymarchonShaun** [comment]
As a workaround for sway, has had any success with `no_focus`? 
I'm trying `no_focus [class="steam"]` right now, but I con't confirm it works yet.

**Comment by DrymarchonShaun** [comment]
Update to my previous comment - as far as I can tell, it works well enough for a temporary solution. I have noticed some weirdness potentially caused by the workaround, but haven't had enough time to know if they are consistently occurring.

**Comment by SuperTux88** [comment]
`no_focus [class="steam"]` didn't make any difference to me, and notifications still block focus :disappointed: ... I'm just disabling notifications in steam instead, but I reenabled them to test if the `no_focus` works.

**Comment by Sid127** [comment]
This seems to be fixed in the current steam client

**Comment by J-Lowrance** [comment]
I'm still having this issue with the Steam Client Beta, using KDE on X. Seems to be related to #9716

**Comment by ibrokemypie** [comment]
Having this issue on i3wm with steam version 1696019606 (on nvidia)

**Comment by MisterSheeple** [comment]
Oddly this same thing happened to me when I used X11. When I switched to Wayland, it didn't do it anymore, but recently I upgraded to Ubuntu 23.10 and it now does this on Wayland.

**Comment by JangXa** [comment]
Fixed for me by adding following rule to my hypr.conf:

`windowrulev2 = noinitialfocus, class:(^steam.*) `

**Comment by Chais** [comment]
> Fixed for me by adding following rule to my hypr.conf:
> 
> `windowrulev2 = noinitialfocus, title:(^notificationtoasts.*)`

Sadly I can't confirm this working.
